File: 0x08-python-more_classes/3-rectangle.py
#!/usr/bin/python3
"""defines class Rectangle"""


class Rectangle:
    """
    class Rectangle with private instance attributes width & height
    attributes have properties setter and getter
    setter properties have type and value validation with exceptions
    """
    def __init__(self, width=0, height=0):
        """
        instance initialization method

        Args:
        width (int)
        height (int)
        """
        self.width = width
        self.height = height

    def __str__(self):
        """prints the rectangle with the character #"""
        if self.__width == 0 or self.__height == 0:
            return ("")
        else:
            str_rectangle = ""
            for i in range(self.__height):
                for j in range(self.__width):
                    str_rectangle += '#'
                if i < (self.__height - 1):
                    str_rectangle += '\n'
            return (str_rectangle)

    @property
    def width(self):
        """get the width of the rectangle"""
        return (self.__width)

    @width.setter
    def width(self, value: int):
        if not (isinstance(value, int)):
            raise TypeError('width must be an integer')
        if value < 0:
            raise ValueError('width must be >= 0')
        self.__width = value

    @property
    def height(self):
        """get the height of the rectangle"""
        return (self.__height)

    @height.setter
    def height(self, value: int):
        if not (isinstance(value, int)):
            raise TypeError('height must be an integer')
        if value < 0:
            raise ValueError('height must be >= 0')
        self.__height = value

    def area(self):
        """returns the area of the rectangle"""
        return (self.__height * self.__width)

    def perimeter(self):
        """returns the perimeter of the rectangle"""
        if self.__width == 0 or self.__height == 0:
            return (0)
        else:
            return ((self.__height + self.__width) * 2)

File: 0x13-javascript_objects_scopes_closures/3-rectangle.js
#!/usr/bin/node
/* class Rectangle that defines a rectangle
* the constructor takes 2 arguments w and h
* - initializes the instance attribute width with the value of w
* - initializes the instance attribute height with the value of h
* - if w or h is equal to 0 or not a positive integer, it creates an empty object
* instance method print prints the rectangle using the character 'X'
*/
class Rectangle {
  constructor (w, h) {
    if (w > 1 & h > 1) {
      this.width = w;
      this.height = h;
    }
  }

  print () {
    let k = '';
    for (let i = 0; i < this.height; i++) {
      k = '';
      for (let j = 0; j < this.width; j++) {
        k += 'X';
      }
      console.log(k);
    }
  }
}
module.exports = Rectangle;
